File: files/elastic-solvers/direct_iteration.py
"""Direct perturbative spectral iteration. GPL-3.0-or-later.
Based on the reference-stiffness split in the group's C/Fortran solvers
and Bhattacharyya et al., CICP 11 (2012), 726–738, Eqs. 2.6–2.16.
This is a small, independently written 2-D teaching adaptation.
"""
import numpy as np
from elasticity_common import prepare, stress, diagnostics, result, plate_problem, plot_solution


def solve_direct(stiffness, eigenstrain, reference, mean_strain=None,
                 relaxation=0.8, tolerance=1e-8, max_iterations=4000):
    grid, mean_strain, scale = prepare(stiffness,eigenstrain,mean_strain,reference,
                                      relaxation,tolerance,max_iterations)
    displacement = np.zeros(eigenstrain.shape[:2]+(2,))
    contrast = stiffness-reference
    history=[]
    for iterator in range(max_iterations+1):
        strain = mean_strain+grid.strain(displacement)
        sigma, energy_density, residual = diagnostics(grid,stiffness,eigenstrain,strain,scale)
        history.append([iterator,residual,float(energy_density.mean())])
        if residual < tolerance:
            return result(grid,stiffness,eigenstrain,strain,displacement,scale,history)
        # Move all heterogeneous terms to the RHS of a homogeneous problem:
        # Cref:epsilon - P = C:(epsilon-eigenstrain).
        polarization = stress(stiffness,eigenstrain)-stress(contrast,strain)
        candidate = grid.response(reference,polarization)
        displacement += relaxation*(candidate-displacement)
        if not np.isfinite(residual) or residual > 1e12:
            break
    raise RuntimeError(f'Direct iteration did not converge; residual={residual:.3e}. '
                       'Try a better reference, lower contrast or smaller relaxation.')


if __name__ == '__main__':
    stiffness,eigenstrain,reference,phase = plate_problem()
    solution=solve_direct(stiffness,eigenstrain,reference)
    print(f"Energy={solution['energy']:.9g}; residual={solution['residual']:.3g}")
    plot_solution(solution,phase).savefig('direct-iteration.png',dpi=180)

File: files/elastic-solvers/DirectIteration.jl
# Direct perturbative spectral iteration. GPL-3.0-or-later.
isdefined(@__MODULE__, :ElasticityCommon) || include("ElasticityCommon.jl")
using .ElasticityCommon, Printf, Statistics

function solve_direct(stiffness,eigenstrain,reference;mean_strain=nothing,
                      relaxation=0.8,tolerance=1e-8,max_iterations=4000)
    grid,mean_tensor,scale = prepare(stiffness,eigenstrain,mean_strain,reference,relaxation,tolerance,max_iterations)
    displacement = zeros(size(eigenstrain,1),size(eigenstrain,2),2)
    contrast = stiffness .- reshape(reference,1,1,2,2,2,2)
    history = Vector{Vector{Float64}}()
    for iterator in 0:max_iterations
        total_strain = reshape(mean_tensor,1,1,2,2) .+ ElasticityCommon.strain(grid,displacement)
        sigma,energy_density,residual = diagnostics(grid,stiffness,eigenstrain,total_strain,scale)
        push!(history,[iterator,residual,mean(energy_density)])
        residual < tolerance && return package_result(grid,stiffness,eigenstrain,total_strain,displacement,scale,history)
        polarization = stress(stiffness,eigenstrain)-stress(contrast,total_strain)
        candidate = ElasticityCommon.response(grid,reference,polarization)
        displacement .+= relaxation.*(candidate-displacement)
        isfinite(residual) && residual < 1e12 || break
    end
    error("Direct iteration did not converge. Try a better reference or smaller relaxation.")
end

if abspath(PROGRAM_FILE) == @__FILE__
    stiffness,eigenstrain,reference,phase = plate_problem()
    solution = solve_direct(stiffness,eigenstrain,reference)
    @printf "Energy = %.9g; residual = %.3g\n" solution.energy solution.residual
end
